File: Daily-Streak/Day-10_Minimum-ASCII-Delete-Sum-for-Two-Strings/Solution.py
class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        n, m = len(s1), len(s2)

        # dp[i][j] = minimum ASCII delete sum
        dp = [[0] * (m + 1) for _ in range(n + 1)]

        # Base case: s2 exhausted
        for i in range(n - 1, -1, -1):
            dp[i][m] = dp[i + 1][m] + ord(s1[i])

        # Base case: s1 exhausted
        for j in range(m - 1, -1, -1):
            dp[n][j] = dp[n][j + 1] + ord(s2[j])

        # Bottom-up DP
        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                if s1[i] == s2[j]:
                    dp[i][j] = dp[i + 1][j + 1]
                else:
                    dp[i][j] = min(
                        ord(s1[i]) + dp[i + 1][j],
                        ord(s2[j]) + dp[i][j + 1]
                    )

        return dp[0][0]

        """
        ===========================
        ❌ NAIVE IDEA (COMMENTED)
        ===========================
        - Try all delete combinations
        - Exponential complexity
        - Not feasible

        ===========================
        🧪 STANDALONE VERSION (COMMENTED)
        ===========================
        def main():
            s1 = "sea"
            s2 = "eat"

            sol = Solution()
            print(sol.minimumDeleteSum(s1, s2))

        if __name__ == "__main__":
            main()
        """

File: Daily-Streak/Day-10_Minimum-ASCII-Delete-Sum-for-Two-Strings/Solution.cpp
class Solution {
public:
    int minimumDeleteSum(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();

        // dp[i][j] = minimum ASCII delete sum
        // to make s1[i:] and s2[j:] equal
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));

        // Base case: s2 exhausted
        for (int i = n - 1; i >= 0; i--) {
            dp[i][m] = dp[i + 1][m] + s1[i];
        }

        // Base case: s1 exhausted
        for (int j = m - 1; j >= 0; j--) {
            dp[n][j] = dp[n][j + 1] + s2[j];
        }

        // Fill DP table bottom-up
        for (int i = n - 1; i >= 0; i--) {
            for (int j = m - 1; j >= 0; j--) {
                if (s1[i] == s2[j]) {
                    dp[i][j] = dp[i + 1][j + 1];
                } else {
                    dp[i][j] = min(
                        s1[i] + dp[i + 1][j],
                        s2[j] + dp[i][j + 1]
                    );
                }
            }
        }

        return dp[0][0];
    }

    /*
    ===========================
    ❌ NAIVE IDEA (COMMENTED)
    ===========================
    - Try deleting characters recursively
    - Compare all possible strings
    - Exponential time → impossible

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    #include <bits/stdc++.h>
    using namespace std;

    int main() {
        string s1 = "sea";
        string s2 = "eat";

        Solution sol;
        cout << sol.minimumDeleteSum(s1, s2) << endl;
        return 0;
    }
    */
};
